mlflow/utils/autologging_utils.py

```diff
@@ -275,3 +282,202 @@ def batch_metrics_logger(run_id):
+    integrations should be decorated with this wrapper.**
+    """
+
+    AUTOLOGGING_INTEGRATIONS[name] = {}
+
+    def wrapper(_autolog):
+        def autolog(*args, **kwargs):
+            AUTOLOGGING_INTEGRATIONS[name] = kwargs
+            _autolog(**kwargs)
+
+        wrapped_autolog = functools.wraps(_autolog)(autolog)
+        return wrapped_autolog
+
+    return wrapper
+
+
+def safe_patch(autologging_integration, destination, function_name, function):
+    """
+    Patches the specified `function_name` on the specified `destination` class for autologging
+    purposes, replacing its implementation with an error-safe copy of the specified `function`.
+
+    :param autologging_integration: The name of the autologging integration associated with the
+                                    patch.
+    :param destination: The Python class on which the patch function is being defined.
+    :param function_name: The name of the function to patch on the specified `destination` class.
+    :param function: The patch function to apply. The first argument to this function should be
+                     reserved for an `original` method argument representing the underlying /
+                     original function. Subsequent arguments should be identical to those of the
+                     original function being patched.
+    """
+
+    def patched_train(*args, **kwargs):
+        preexisting_run = mlflow.active_run()
+        original_result = None
+        called_original = False
+        failed_during_original = False
+
+        def call_original(*og_args, **og_kwargs):
+            original = gorilla.get_original_attribute(destination, function_name)
+
+            if _is_testing():
+                _validate_args(args, kwargs, og_args, og_kwargs)
+
+            def wrapped_original(*args, **kwargs):
+                try:
+                    called_original = True
+                    nonlocal original_result
+                    original_result = original(*args, **kwargs)
+                    return original_result
+                except Exception:
+                    nonlocal failed_during_original
+                    failed_during_original = True
+                    raise
+
+            return wrapped_original(*og_args, **og_kwargs)
+
+        original = gorilla.get_original_attribute(destination, function_name)
+        call_original = functools.wraps(original)(call_original)
+        call_original.__signature__ = inspect.signature(original)
+
+        config = AUTOLOGGING_INTEGRATIONS[autologging_integration]
+        if config.get("disable", False):
+            return original(*args, **kwargs)
+
+        try:
+            return function(call_original, *args, **kwargs)
+        except Exception as e:
+            if _is_testing():
+                raise
+
+            if not preexisting_run and mlflow.active_run():
+                try_mlflow_log(mlflow.end_run, RunStatus.to_string(RunStatus.FAILED))
```

Comment:
Here, we attempt to clean up any dangling runs. As pointed out by @aarondav , it might be better to structure this as part of the patch API via an `on_exception` method defined on a patch class; this way, it's very clear to developers that teardown code is necessary & allows them to augment this teardown code with additional logic.